iree/compiler/Codegen/Common/WorkGroupSwizzle.cpp

```diff
@@ -0,0 +1,113 @@
+// Copyright 2022 The IREE Authors
+//
+// Licensed under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+#include "iree/compiler/Codegen/PassDetail.h"
+#include "iree/compiler/Codegen/Passes.h"
+
+namespace mlir {
+namespace iree_compiler {
+
+/// This function implements the following swizzling logic
+/// void getTiledId2(unsigned x, unsigned y, unsigned* tiledx,
+///                 unsigned* tiledy) {
+///  unsigned t_tiledx = (x + (y % tile) * grid_size_x) / tile;
+///  unsigned t_tiledy = (y / tile) * tile +
+///      (x + (y % tile) * grid_size_x) % tile;
+///  bool c = grid_size_y % tile != 0 &&
+///      ((y / tile) * tile + tile) > grid_size_y;
+///  *tiledx = c ? x : t_tiledx;
+///  *tiledy = c ? y : t_tiledy;
+/// }
+static void makeSwizzledId(Location loc, OpBuilder b, Value workgroupIdX,
```

Comment:
Not required for this PR, but you could make this swizzling a callback and make the core transformation a utility function that takes the call back to get the swizzling. You can use this method as the callback for your use case as a pass in GPU.

Reply:
Could yo add a TODO here that captures that this function could be made a call back and that the core logic in this pass can be made into a utility function?

Reply:
Done.

Reply:
Actually the TODO should be `TODO : Make this a callback and the core functionality in the pass a utility function`.